witness/golang/omniwitness/usbarmory/storage/slots/journal.go

```diff
@@ -0,0 +1,258 @@
+func (e *entry) Size() int {
+	return entryHeaderSize + len(e.Data)
+}
+
+// OpenJournal returns a new journal structure for interacting with a journal stored in the
+// [start, start+length) range of blocks accessible via dev.
+func OpenJournal(dev BlockReaderWriter, start, length uint) (*Journal, error) {
+	j := &Journal{
+		dev:           dev,
+		start:         start,
+		length:        length,
+		maxWriteBytes: (length*dev.BlockSize())/3 - entryHeaderSize,
+	}
+
+	if err := j.init(); err != nil {
+		return nil, err
+	}
+
+	return j, nil
+}
+
+// Data returns the application data from the most recent valid entry in the
+// journal, along with the entry's revision number.
+func (j *Journal) Data() ([]byte, uint32) {
+	return j.current.Data, j.current.Revision
+}
+
+// Update creates a new entry record in the journal for the provided data.
+// The new record's revision will be one greater than the previous record (or 1
+// if no previous record exists).
+func (j *Journal) Update(data []byte) error {
+	if l := len(data); l > int(j.maxWriteBytes) {
+		return fmt.Errorf("attemping to write %d bytes, larger than the max permitted in this journal (%d bytes)", l, j.maxWriteBytes)
+	}
+	e := &entry{
+		Magic:      [4]byte{magic0[0], magic0[1], magic0[2], magic0[3]},
+		Revision:   j.current.Revision + 1,
+		DataLen:    uint64(len(data)),
+		DataSHA256: sha256.Sum256(data),
+		Data:       data,
+	}
+
+	cap := (j.length - j.nextBlock) * j.dev.BlockSize()
+	if cap < uint(e.Size()) {
+		// The record won't fit in the remaining space, so wrap around and write at the beginning.
+		j.nextBlock = j.start
+	}
+	// TODO(al): consider making this more "streamy".
+	buf := &bytes.Buffer{}
+	if err := marshalEntry(e, buf); err != nil {
+		return fmt.Errorf("failed to marshal entry: %v", err)
+	}
+	return j.dev.WriteBlocks(j.nextBlock, padToBlock(buf.Bytes(), j.dev.BlockSize()))
+}
+
+// Init scans the journal to figure out the latest valid record, if any.
+func (j *Journal) init() error {
+	j.nextBlock = j.start
+	// Start where all good stories do: at the beginning!
+	lba := j.start
+	var lastEntry *entry
+	for lba < j.start+j.length {
+		br := newBlockReader(j.dev, lba)
+		e, err := unmarshalEntry(br)
+		if err != nil {
+			if lastEntry != nil {
+				// We already found the lastet record in the journal, so we're done.
+				j.current = *lastEntry
+				j.nextBlock = lba
+				return nil
+			}
+			// Since we couldn't unmarshal the entry struct we'll have to fall back to
+			// scanning all blocks.
+			lba++
+			continue
+		}
+		if lastEntry == nil || e.Revision > lastEntry.Revision {
+			lastEntry = e
+			lba = (br.pos-1)/br.dev.BlockSize() + 1
+			continue
+		} else if e.Revision < lastEntry.Revision {
+			// We've found an older revision following a newer one, so we're done
+			j.current = *lastEntry
+			return nil
+		} else {
+			return fmt.Errorf("journal is corrupt - found two entries with the same revision (%d)", e.Revision)
+		}
+	}
+	return nil
```

Comment:
I think you need to do something here with assigning `lastEntry` to `j.current` ? There is a case where you update `lastEntry` above, and if that perfectly aligned with the end of the journal then it looks like you'll then snap out of the loop and then drop it on the floor.

It would also be good to be explicit about how `j.current` is initialized in the case of an empty journal. I hinted at it on the comment for `Data()` above, but it's an important case to make sure we get right.

Reply:
I _think_ it was ok, but I've jiggered thing around a bit to try to make it clearer - now we only assign to j.current etc. in one place (at the bottom of the func) - early returns became `break`s etc.